torchtext/csrc/vectors.cpp

```diff
@@ -70,6 +76,45 @@ struct Vectors : torch::CustomClassHolder {
   int64_t __len__() { return stovec_.size(); }
 };
 
+VectorsStates _set_vectors_states(const c10::intrusive_ptr<Vectors> &self) {
+  std::vector<int64_t> integers;
+  std::vector<std::string> strings = self->tokens_;
+  std::vector<torch::Tensor> tensors{self->vectors_, self->unk_tensor_};
+
+  VectorsStates states =
+      std::make_tuple(self->version_str_, std::move(integers),
+                      std::move(strings), std::move(tensors));
+
+  return states;
+}
+
+c10::intrusive_ptr<Vectors> _get_vectors_from_states(VectorsStates states) {
+  auto state_size = std::tuple_size<decltype(states)>::value;
+  if (state_size != 4) {
+    throw std::runtime_error(
+        "Expected deserialized Vectors to have 4 states but found only " +
+        std::to_string(state_size) + " states.");
+  }
```

Comment:
Hi @Nayef211 Could you help me understand this? As far as I understand, `std::tuple_size<decltype(states)>` is resolved at compile time so `state_size` is alway 4.

Reply:
or is there a case that Python can pass tuple with different size?

Reply:
Hi @mthrok. I don't think there's any Python involved in this part of the code. The `_set_vectors_states ` states function always creates a tuple of size 4 when a `vector` is serialized, and the `_get_vectors_from_states ` always expects 4 states during deserialization. `std::tuple_size<decltype(states)>` actually allows us to detect the number of items inside of states before deserializing. The reason this is done is for backward compatibility since in the future we may change the number of items in the tuple that we serialize in the `states` tuple. If the number of items has indeed changed, then we can return an error message to the user so they understand why they can't deserialize their serialized vector.

Lmk if that helps!

Reply:
@Nayef211 Grad to hear back from you. Hope you are doing well.

I understand the purpose of the check, and what I wonder is does the check really prevent the expected undesirable situation? The function does not seem to implement runtime polymorphism, so the questions I am asking myself are;
1. When passing a different type of object (say, tuple with different number) from Python side, does the program execution reach this function? Does PyBind11 find a way to pass a different type to this function?
2. If PyBind11 manages to pass the wrong object to this function, (say it uses some sort of `reinterpret_cast`-ing), isn't the object still represented as the correct `VectorsStates` ?

So the question to you is "Did you test such undeniable case and verified that the check here detect the situation"? (By asking this, I am not accusing you of anything, I would like to check it with you before I do the test by myself.)